File: order.service/src/main/java/br/com/edu/order/domain/Item.java
/* (C)2024 */
package br.com.edu.order.domain;

import br.com.edu.order.domain.errors.InvalidPriceException;
import br.com.edu.order.domain.errors.InvalidQuantityException;
import jakarta.persistence.*;
import java.math.BigDecimal;
import java.util.UUID;
import lombok.Data;

@Data
@Entity
@Table(name = "Items")
public class Item {

    @Id
    @Column(columnDefinition = "uuid")
    private UUID id;

    @Column(name = "order_id", columnDefinition = "uuid")
    private UUID orderId;

    @Column(name = "product")
    private String product;

    @Column(name = "price")
    private BigDecimal price;

    @Column(name = "quantity")
    private int quantity;

    public Item() {
        this.id = UUID.randomUUID();
        this.price = BigDecimal.ZERO;
        this.quantity = 0;
    }

    public void setPrice(final BigDecimal price) {
        if (price.compareTo(BigDecimal.ZERO) < 1) {
            throw new InvalidPriceException(price);
        }
        this.price = price;
    }

    public void setQuantity(int quantity) {
        if (quantity <= 0) {
            throw new InvalidQuantityException(quantity);
        }
        this.quantity = quantity;
    }

    public BigDecimal totalItem() {
        return price.multiply(BigDecimal.valueOf(quantity));
    }
}

File: order.service/order-service-usecase/src/main/kotlin/br/com/edu/order/domain/Item.kt
package br.com.edu.order.domain

import br.com.edu.order.domain.errors.InvalidPriceException
import br.com.edu.order.domain.errors.InvalidQuantityException
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.Id
import jakarta.persistence.Table
import java.math.BigDecimal
import java.util.*

@Entity
@Table(name = "Items")
class Item(

    @Id
    @Column(columnDefinition = "uuid")
    var id: UUID = UUID.randomUUID(),

    @Column(name = "order_id", columnDefinition = "uuid")
    var orderId: UUID? = null,

    @Column(name = "product")
    var product: String? = null

) {

    @Column(name = "quantity")
    var quantity: Int = 0
        set(value) {
            if (value <= 0) {
                throw InvalidQuantityException(value)
            }
            field = value
        }

    @Column(name = "price")
    var price: BigDecimal = BigDecimal.ZERO
        set(value) {
            if (value.compareTo(BigDecimal.ZERO) < 1) {
                throw InvalidPriceException(value)
            }
            field = value
        }

    fun totalItem(): BigDecimal {
        return price.multiply(BigDecimal.valueOf(quantity.toLong()))
    }
}
